core/input/input.cpp

```diff
@@ -417,9 +557,26 @@ float Input::get_joy_axis(int p_device, JoyAxis p_axis) const {
 	}
 }
 
-String Input::get_joy_name(int p_idx) {
+String Input::get_joy_name(int p_device) {
 	_THREAD_SAFE_METHOD_
-	return joy_names[p_idx].name;
+	return joy_names[p_device].name;
+}
+
+Input::JoyScheme Input::get_joy_scheme(int p_device) {
+	String joy_name = get_joy_name(p_device).to_lower();
+
+	if (joy_name.contains("nintendo") || joy_name.contains("joy-con")) {
+		return JOY_SCHEME_NINTENDO;
+	}
+	if (joy_name.contains("sony") || joy_name.contains("playstation") || joy_name.contains("dualshock") ||
+			joy_name.contains("ps1") || joy_name.contains("ps2") || joy_name.contains("ps3") || joy_name.contains("ps4") || joy_name.contains("ps5")) {
```

Comment:
What names do official controllers use. Is this "just to be sure"?

Reply:
I mostly got those names from the controller DB file: https://github.com/godotengine/godot/blob/master/core/input/gamecontrollerdb.txt